File: fase-4/voz.py
# Importa os — acessa variáveis do sistema operacional
import os

# Importa sounddevice — captura áudio do microfone
import sounddevice as sd

# Importa write — salva o áudio capturado em arquivo WAV
from scipy.io.wavfile import write

# Importa numpy — operações matemáticas em arrays de áudio
import numpy as np

# Importa pyttsx3 — converte texto em fala (TTS) offline
import pyttsx3

# Importa Groq — conecta com a API do Groq
from groq import Groq

# Importa load_dotenv — carrega o .env
from dotenv import load_dotenv

# Importa Path — trabalha com caminhos de arquivos
from pathlib import Path

# Carrega o .env da raiz do projeto
load_dotenv(dotenv_path=Path(__file__).parent.parent / ".env")

# Cria o cliente Groq
client = Groq(api_key=os.getenv("GROQ_API_KEY"))

# Histórico da conversa
historico = [
    {
        "role": "system",
        "content": "Você é o Jarvis, assistente pessoal do Gabriel. Seja direto e inteligente. Responda de forma curta e objetiva pois sua resposta será lida em voz alta. Sempre chame o usuário de 'senhor Gabriel'."
    }
]

def transcrever_groq(caminho_audio):
    """Envia o áudio pro Groq e retorna o texto transcrito"""
    with open(caminho_audio, "rb") as f:
        resposta = client.audio.transcriptions.create(
            file=f,
            model="whisper-large-v3",
            response_format="json"
        )
    return resposta.text

def falar(texto):
    """Converte texto em fala — TTS"""
    print(f"Jarvis: {texto}")
    try:
         # Reinicia o motor a cada fala — resolve o travamento no Windows
        motor = pyttsx3.init()
        motor.setProperty("rate", 150)
        motor.say(texto)
        motor.runAndWait()
        motor.stop()
    except Exception as e:
        print(f"Erro no TTS: {e}")

def ouvir():
    """Captura áudio do microfone com detecção de silêncio"""
    fs = 16000           # frequência de amostragem
    chunk = 1024         # pedaços de áudio analisados por vez
    silencio_limite = 0.05  # volume mínimo pra considerar voz
    silencio_max = 2     # segundos de silêncio pra parar de gravar

    print("🎤 Ouvindo... (fale agora)")

    gravando = []        # acumula os pedaços de áudio
    silencio = 0         # contador de silêncio
    falando = False      # flag — detectou voz?

    with sd.InputStream(samplerate=fs, channels=1, dtype='float32') as stream:
        while True:
            pedaco, _ = stream.read(chunk)
            volume = np.abs(pedaco).mean()

            if volume > silencio_limite:
                falando = True
                silencio = 0
                gravando.append(pedaco)
            elif falando:
                silencio += chunk / fs
                gravando.append(pedaco)
                if silencio >= silencio_max:
                    break

    # Junta os pedaços e salva o arquivo WAV
    audio = np.concatenate(gravando, axis=0)
    write("audio.wav", fs, audio)

    texto = transcrever_groq("audio.wav")
    texto = texto.strip()
    print(f"Você disse: {texto}")
    return texto if texto else None

def perguntar_jarvis(texto):
    """Envia o texto pro Groq e retorna a resposta"""
    historico.append({"role": "user", "content": texto})

    resposta = client.chat.completions.create(
        model="llama-3.1-8b-instant",
        messages=historico
    )

    mensagem = resposta.choices[0].message.content
    historico.append({"role": "assistant", "content": mensagem})
    return mensagem

# ── LOOP PRINCIPAL ─────────────────────────────────────
print("Jarvis online! Fale algo ou diga 'sair' para encerrar.\n")
falar("Olá senhor Gabriel, estou online e pronto para ajudá-lo.")

while True:
    entrada = ouvir()

    if entrada is None:
        falar("Não entendi, pode repetir?")
        continue

    if "sair" in entrada.lower():
        falar("Até logo, senhor Gabriel.")
        break

    resposta = perguntar_jarvis(entrada)
    falar(resposta)

File: fase-4/voz.js
// Importa Groq — SDK oficial para conectar com a API do Groq
import Groq from "groq-sdk"

// Importa dotenv — carrega as variáveis do .env
import dotenv from "dotenv"

// Importa fileURLToPath, dirname, join — reconstrói o __dirname no ES Module
import { fileURLToPath } from "url"
import { dirname, join } from "path"

// Importa fs — lê e escreve arquivos no sistema
import { createReadStream, createWriteStream } from "fs"

// Importa say — converte texto em fala (TTS)
import say from "say"

// Importa Microphone — captura áudio do microfone no Windows
import Microphone from "node-microphone"

// Reconstrói o __dirname
const __dirname = dirname(fileURLToPath(import.meta.url))

// Carrega o .env
dotenv.config({ path: join(__dirname, "..", ".env") })

// Cria o cliente Groq
const client = new Groq({ apiKey: process.env.GROQ_API_KEY })

// Histórico da conversa
const historico = [
    {
        role: "system",
        content: "Você é o Jarvis, assistente pessoal do Gabriel. Seja direto e inteligente. Responda de forma curta e objetiva pois sua resposta será lida em voz alta. Sempre chame o usuário de 'senhor Gabriel'."
    }
]

// ── FUNÇÃO TTS ─────────────────────────────────────────
function falar(texto) {
    return new Promise((resolve) => {
        console.log(`Jarvis: ${texto}`)
        say.speak(texto, null, 1.0, () => resolve())
    })
}

// ── FUNÇÃO STT ─────────────────────────────────────────
function ouvir() {
    return new Promise((resolve) => {
        const arquivoAudio = join(__dirname, "audio.wav")
        const mic = new Microphone()
        const stream = mic.startRecording()
        const writer = createWriteStream(arquivoAudio)

        console.log("🎤 Ouvindo... (aguarde 5 segundos)")

        // Escreve o áudio no arquivo
        stream.pipe(writer)

        // Para após 5 segundos
        setTimeout(async () => {
            mic.stopRecording()
            writer.end()

            // Aguarda o arquivo ser salvo
            writer.on("finish", async () => {
                try {
                    const audioStream = createReadStream(arquivoAudio)
                    const resposta = await client.audio.transcriptions.create({
                        file: audioStream,
                        model: "whisper-large-v3",
                        response_format: "json",
                        language: "pt"
                    })
                    const texto = resposta.text.trim()
                    console.log(`Você disse: ${texto}`)
                    resolve(texto || null)
                } catch (erro) {
                    console.log(`Erro na transcrição: ${erro.message}`)
                    resolve(null)
                }
            })
        }, 5000)
    })
}

// ── FUNÇÃO JARVIS ──────────────────────────────────────
async function perguntarJarvis(texto) {
    historico.push({ role: "user", content: texto })

    const resposta = await client.chat.completions.create({
        model: "llama-3.1-8b-instant",
        messages: historico
    })

    const mensagem = resposta.choices[0].message.content
    historico.push({ role: "assistant", content: mensagem })
    return mensagem
}

// ── LOOP PRINCIPAL ─────────────────────────────────────
async function main() {
    console.log("Jarvis online! Fale algo ou diga 'sair' para encerrar.\n")
    await falar("Olá senhor Gabriel, estou online e pronto para ajudá-lo.")

    while (true) {
        const entrada = await ouvir()

        if (!entrada) {
            await falar("Não entendi, pode repetir?")
            continue
        }

        if (entrada.toLowerCase().includes("sair")) {
            await falar("Até logo, senhor Gabriel.")
            break
        }

        const resposta = await perguntarJarvis(entrada)
        await falar(resposta)
    }
}

main()